README.md

```diff
@@ -38,6 +38,72 @@ If you want to test what happens when the HTTP server goes down, use `Bypass.dow
 TCP socket and `Bypass.up/1` to start listening on the same port again. Both functions block until
 the socket updates its state.
 
+### Expect Functions
+
+You can take any of the following approaches:
+* `expect/2` or `expect_once/2` to install a generic function that all calls to bypass will use
+* `expect/4` and/or `expect_once/4` to install specific routes (method, path combinations)
+* a combination of the above, where the routes will be used first, and then the generic version
+  will be used as default
+
+#### expect/2 (bypass_instance, function)
+
+Must be called at least once, and it will error if not.  As an alternative, `nil` can be passed
+in place of a function to indicate that it should never be called (and it will error if it is).
+
+```elixir
+  Bypass.expect bypass, fn conn ->
+    assert "/1.1/statuses/update.json" == conn.request_path
+    assert "POST" == conn.method
+    Plug.Conn.resp(conn, 429, ~s<{"errors": [{"code": 88, "message": "Rate limit exceeded"}]}>)
+  end
+```
+
+#### expect_once/2 (bypass_instance, function)
+
+Must be called exactly once, and it will error if not. As an alternative, `nil` can be passed
+in place of a function to indicate that it should never be called (and it will error if it is)
+
+```elixir
+  Bypass.expect_once bypass, fn conn ->
+    assert "/1.1/statuses/update.json" == conn.request_path
+    assert "POST" == conn.method
+    Plug.Conn.resp(conn, 429, ~s<{"errors": [{"code": 88, "message": "Rate limit exceeded"}]}>)
+  end
+```
+
+#### expect/4 (bypass_instance, method | [method1, method2...], path | [path1, path2...], function)
+
+Must be called at least once, and it will error if not.  As an alternative, `nil` can be passed
+in place of a function to indicate that it should never be called (and it will error if it is).
+
+The methods can be passed as a binary or a `List` of binaries, and the paths can also be passed in
+that way.  All combinations will be required to be called.
```

Comment:
> All combinations will be required to be called.

In other words, there should be at least one request for every `{method, path}` specified in order for bypass to succeed?

This is also more complex than it should be. I know this comes from what I wrote in my proposal:

> The second form allows installing multiple handlers so long as each one has a unique method-route combination, each of which has to be invoked at least once for the test to pass.

But now that I think of it, it is not necessary to support a list of methods or paths.

Let's look at an example:

```elixir
Bypass.expect bypass, ["GET", "POST"], "/1.1/statuses/update.json", fn conn ->
  Agent.get_and_update(AgentModule, fn step_no -> {step_no, step_no+1} end)
  Plug.Conn.resp(conn, 429, ~s<{"errors": [{"code": 88, "message": "Rate limit exceeded"}]}>)
end
```

It reads as if one of `GET` or `POST` requests were sufficient to satisfy the expectation. It is more likely that different methods will require different handler functions. Even when the handler function is the same, the following code looks much clearer in its intent:

```elixir
Enum.each(["GET", "POST"], fn method ->
  Bypass.expect bypass, method, "/1.1/statuses/update.json", fn conn ->
    Agent.get_and_update(AgentModule, fn step_no -> {step_no, step_no+1} end)
    Plug.Conn.resp(conn, 429, ~s<{"errors": [{"code": 88, "message": "Rate limit exceeded"}]}>)
  end
end)
```

In the unusual case when multiple methods and paths should use the same handler function, I would prefer unrolling the loop:

```elixir
handler_fn = fn conn ->
  Agent.get_and_update(AgentModule, fn step_no -> {step_no, step_no+1} end)
  Plug.Conn.resp(conn, 429, ~s<{"errors": [{"code": 88, "message": "Rate limit exceeded"}]}>)
end

Bypass.expect bypass, "GET", "/1.1/statuses/status.json", handler_fn
Bypass.expect bypass, "POST", "/1.1/statuses/update.json", handler_fn
```

The idea is simple: don't complicate the API if the same result can be achieved using built-in functionality.